### ISSUE TITLE
Custom tensorflow or pytorch based model

### ISSUE THREAD

**Issue by joel5638** [davisking/dlib#1988]
Hi @davisking  I understand your dlib models are  regression tree based. Can you help me or give me any reference with any tensor flow based process using faster rcnn or any model?

**Comment by arrufat** [comment]
When you are talking about the regression tree based models, I am assuming you are referring to dlib's [shape predictor](http://dlib.net/imaging.html#shape_predictor), dlib allows you to train deep neural networks, as well. I think I might be misunderstanding you.

**Comment by joel5638** [comment]
@arrufat   how does dlib allow in training a neural network? Please give me a reference il read through. Im trying to create a custom landmark detector using using tensorflow. But im confused.

**Comment by arrufat** [comment]
You should definitely read:

- http://dlib.net/dnn_introduction_ex.cpp.html
- http://dlib.net/dnn_introduction2_ex.cpp.html

Also, to train a landmark detector using neural you can use this loss layer:
- http://dlib.net/dlib/dnn/loss_abstract.h.html#loss_mean_squared_per_channel_and_pixel_

Its use case is to estimate the heatmap of landmarks, similar to this paper:
- https://arxiv.org/abs/1804.06208

**Comment by joel5638** [comment]
@arrufat wonderful. 
One last question. I see them in C++. Im unaware of it. Can i find anything in Python?

**Comment by arrufat** [comment]
Not for training deep neural networks.

**Comment by joel5638** [comment]
@arrufat oh. So we cant train a network to detect facial landmarks with python?

**Comment by davisking** [comment]
There is a great deal of documentation at http://dlib.net/imaging.html#shape_predictor for how to train, including in Python.

**Comment by joel5638** [comment]
@davisking thank you for the references. One thing i want to know is what is the CNN network u used in the shape predictor? 
Can we use our own network too? Like mobilenet and apply regression tree over it?

**Comment by davisking** [comment]
The shape predictor is not a CNN.

**Comment by joel5638** [comment]
@davisking So the shape predictors are bascially built over the regression tree completely and not on any CNN, am I right?

**Comment by davisking** [comment]
Right
